src/transform/src/cse/anf.rs

```diff
@@ -121,6 +124,45 @@ impl Bindings {
     ) -> Result<(), crate::TransformError> {
         self.checked_recur_mut(|this| {
             match relation {
+                MirRelationExpr::LetRec {
+                    ids,
+                    values,
+                    body,
+                    max_iters: _,
+                } => {
+                    // Extend this.rebindings with (id, id) pairs for each
+                    // recursive ID before descending into values and body.
+                    this.rebindings.extend(ids.iter().map(|id| {
+                        let new_id = this.id_offset + u64::cast_from(this.bindings.len());
+                        this.id_offset += 1;
+                        (id.clone(), LocalId::new(new_id))
+                    }));
+
+                    // Descend indo body using a fresh Bindings instance.
+                    for value in values.iter_mut() {
+                        let id_offset = this.id_offset + u64::cast_from(this.bindings.len());
+                        // println!("offset (letrec value #{}) = {}", i, id_offset);
+                        let mut anf = Bindings::new(id_offset, this.rebindings.clone());
```

Comment:
We cannot, because in that case for example in the current test we might:

1. Reference a sub-term that is bound within a prior `LetRec` `value` in a subsequent `value` where this term is not visible (see [this case](https://github.com/MaterializeInc/materialize/pull/19287/commits/4ababb1aafabb09dccd22fc96295dad2faf888eb#diff-3f54d77a9e91b2be438d3610c57ff1c855226ccb27b8efb8a52527c828b05192R1455)).
2. Even if we somehow magically try to do the "pull-up" of such terms so they are all moved from the `value` to the enclosing `LetRec` in the same pass, we should not equate ANF two terms that reference a recursive binding that is updated between them (see [this case](https://github.com/MaterializeInc/materialize/pull/19287/commits/4ababb1aafabb09dccd22fc96295dad2faf888eb#diff-3f54d77a9e91b2be438d3610c57ff1c855226ccb27b8efb8a52527c828b05192R1457)).

Reply:
So this means that we can't discover common subexpressions that appear in different bindings of a LetRec? Do I understand correctly that the `# (1) a Filter (#1 > 7) over l0 appears twice.` comment is pointing out a case of this?

What we could instead do is expire bindings whose meanings have changed, similarly to how the inlining in NormalizeLets does, see `expire_offers`. But if the current version also covers the most important cases that we need, then I guess it's fine for now.

Reply:
Recorded the above comment in the issue description.